bip-csfs.mediawiki

```diff
@@ -0,0 +1,124 @@
+# Verify NOP upgrade: To bring both ECDSA and BIP340 signing to bitcoin script, a NOP upgrade path was chosen for <code>OP_CHECKSIGFROMSTACKVERIFY</code>. This necessarily means leaving the 3 arguments on the stack when executing <code>OP_CHECKSIGFROMSTACKVERIFY</code>. Scripts will need to drop or otherwise manage these stack elements.
+# Add/multisig: No concession is made to <code>OP_CHECKMULTISIG</code> or <code>OP_CHECKSIGADD</code> semantics with <code>OP_CHECKSIGFROMSTACK(VERIFY)</code>. In Tapscript, add semantics can be implemented with 1 additional vByte per key (<code>OP_TOALTSTACK OP_CHECKSIGFROMSTACK OP_FROMALTSTACK OP_ADD OP_TOALTSTACK</code>).
+# Splitting R/S on the stack: Implementing split/separate signatures is left as an exercise for future bitcoin upgrades, such as <code>OP_CAT</code>.
+# BIP118-style Taproot internal key: Rather than introducing an additional key type in this change, we suggest implementing OP_INTERNALKEY or separately introducing that key type for all Tapscript signature checking operations in a separate change.
+# Unknown key lengths: The semantics of other signature checking opcodes in their respective script types (legacy, segwit-v0, tapscript-c0) are applied.
+
+==Resource Limits==
+
+These opcodes are treated identically to other signature checking opcodes and
+count against the various sigops limits and budgets in their respective script
+types.
+
+==Motivation==
+
+===LN Symmetry===
+
+When combined with <code>OP_CHECKTEMPLATEVERIFY</code> (BIP119/CTV),
+<code>OP_CHECKSIGFROMSTACK</code> (CSFS) can be used in Lightning Symmetry channels.
+The construction <code>OP_CHECKTEMPLATEVERIFY <pubkey> OP_CHECKSIGFROMSTACK</code> is
+logically equivalent to <code><bip118_pubkey> OP_CHECKSIG</code> and a signature over
+<code>SIGHASH_ALL|SIGHASH_ANYPREVOUTANYSCRIPT</code>. The <code>OP_CHECKSIGFROMSTACK</code>
+construction is 8 vBytes larger.
+
+===Delegation===
+
+Using a script like:
+<code>OP_DUP <pubkey> OP_CHECKSIGFROMSTACK OP_DROP OP_CHECKSIG</code>
+A script can delegate signing to another key.
+
+==Reference Implementation==
+
+A reference implementation is provided in provided here:
+
+https://github.com/brandonblack/bitcoin/commit/5aae0503ceab93101c459748347a111e4a4852c4
+
+==Backward Compatibility==
+
+By constraining the behavior of an OP_SUCCESS opcode and an OP_NOP opcode,
+deployment of the BIP can be done in a backwards compatible, soft-fork manner.
+If anyone were to rely on the OP_SUCCESS behavior of
+<code>OP_SUCCESS204</code>, <code>OP_CHECKSIGFROMSTACK</code> would invalidate
+their spend.
+
+==Deployment==
+
+TBD
+
+==Credits==
+
+Reference implementation was made with reference to the implementation in
```

Comment:
Perhaps this should also credit Jeremy (as he made a mailing list post soliciting feedback for a design in 2021), and also the authors of the proposal he directly references in the ML?